storagemarket/direct_deals_provider.go

```diff
@@ -404,9 +407,60 @@ func (ddp *DirectDealsProvider) execDeal(ctx context.Context, entry *smtypes.Dir
 		ddp.dealLogger.Infow(dealUuid, "index and announce")
 	}
 
+	if entry.Checkpoint < dealcheckpoints.Complete {
+		// The deal has been added to a piece, so just watch the deal sealing state
+		if derr := ddp.watchSealingUpdates(dealUuid, entry.SectorID); derr != nil {
+			return derr
+		}
+		if err := ddp.updateCheckpoint(ctx, entry, dealcheckpoints.Complete); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+// watchSealingUpdates periodically checks the sealing status of the deal,
+// and returns once the deal is active (or boost is shutdown)
+func (ddp *DirectDealsProvider) watchSealingUpdates(dealUuid uuid.UUID, sectorNum abi.SectorNumber) *dealMakingError {
+	var lastSealingState lapi.SectorState
+	checkSealingFinalized := func() bool {
+		// Get the sector status
+		si, err := ddp.sps.SectorsStatus(ddp.ctx, sectorNum, false)
+		if err == nil && si.State != lastSealingState {
+			// Sector status has changed
+			lastSealingState = si.State
+			ddp.dealLogger.Infow(dealUuid, "current sealing state", "state", si.State)
+		}
+
+		return IsFinalSealingState(si.State)
```

Comment:
if `err != nil` I think this will panic, as `si` will be nil.

Reply:
Thanks - still waking up here :)